packages/fast-components-react-base/src/listbox/listbox.tsx

```diff
@@ -0,0 +1,314 @@
+     * of the context menu
+     */
+    private domChildren(): Element[] {
+        return canUseDOM() && this.rootElement.current instanceof HTMLElement
+            ? Array.from(this.rootElement.current.children)
+            : [];
+    }
+
+    /**
+     * Ensure we always validate our internal state on item focus events, otherwise
+     * the component can get out of sync from click events
+     */
+    private handleMenuItemFocus = (e: React.FocusEvent<HTMLElement>): void => {
+        const target: Element = e.currentTarget;
+        const focusIndex: number = this.domChildren().indexOf(target);
+
+        if (this.isDisabledElement(target)) {
+            target.blur();
+            return;
+        }
+
+        if (focusIndex !== this.state.focusIndex && focusIndex !== -1) {
+            this.setFocus(focusIndex, focusIndex > this.state.focusIndex ? 1 : -1);
+        }
+    };
+
+    /**
+     * Sets focus to the nearest focusable element to the supplied focusIndex.
+     * The adjustment controls how the function searches for other focusable elements
+     * if the element at the focusIndex is not focusable. A positive number will search
+     * towards the end of the children array, whereas a negative number will search towards
+     * the beginning of the children array.
+     */
+    private setFocus(focusIndex: number, adjustment: number): void {
+        const children: Element[] = this.domChildren();
+
+        while (inRange(focusIndex, children.length)) {
+            const child: Element = children[focusIndex];
+
+            if (this.isFocusableElement(child)) {
+                child.focus();
+
+                this.setState({
+                    focusIndex,
+                });
+
+                break;
+            }
+
+            focusIndex += adjustment;
+        }
+    }
+
+    /**
+     * Handle the keydown event of the root menu
+     */
+    private handleMenuKeyDown = (e: React.KeyboardEvent<HTMLDivElement>): void => {
+        switch (e.keyCode) {
+            case KeyCodes.arrowDown:
+            case KeyCodes.arrowRight:
+                e.preventDefault();
+                this.setFocus(this.state.focusIndex + 1, 1);
+
+                break;
+
+            case KeyCodes.arrowUp:
+            case KeyCodes.arrowLeft:
+                e.preventDefault();
+                this.setFocus(this.state.focusIndex - 1, -1);
+
+                break;
+
+            case KeyCodes.end:
+                e.preventDefault();
+                this.setFocus(this.domChildren().length - 1, -1);
+
+                break;
+
+            case KeyCodes.home:
+                e.preventDefault();
+                this.setFocus(0, 1);
+
+                break;
+
+            default:
+                if (!e.ctrlKey) {
+                    this.processTypeAhead(e);
+                }
+        }
+    };
+
+    private processTypeAhead = (e: React.KeyboardEvent<HTMLDivElement>): void => {
+        const acceptedChars: string[] = [
```

Comment:
why do we restrict to just these characters?

Reply:
would it help to check valid characters against a given regex?

Reply:
I'm trying this, seems to work (note that I'm not a regex wiz):
        if  (!/^[a-z0-9\s]+$/i.test(newChar)) {
            return;
        }

Reply:
But why would there ever be a case where we would not want for example, "@"?

Reply:
Pulled out validation in the current iteration so we process all keys.  Chris and I talked about maybe adding a prop where a developper could provide a validation function.  Thoughts?